File: examples/nyml_cli.py
#!/usr/bin/env python3
"""NYML CLI: Parse NYML and output JSON mapping or entries document.

Usage:
  python nyml_cli.py [--entries] [--strategy last|first|all] [-o OUTPUT] FILE

Examples:
  python nyml_cli.py config.nyml
  python nyml_cli.py --entries config.nyml
  python nyml_cli.py --strategy=all config.nyml
"""
import argparse
import json
import os
import sys

# Make sure parsers/python is importable
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'parsers', 'python'))
from nyml_parser import parse_nyml, to_mapping


def main():
    parser = argparse.ArgumentParser(description='Parse NYML and output JSON (mapping or ordered entries)')
    parser.add_argument('file', help='NYML file to parse')
    parser.add_argument('--entries', '-e', action='store_true', help='Output ordered entries document (preserve duplicates)')
    parser.add_argument('--strategy', '-s', choices=['last', 'first', 'all'], default='last', help='Duplicate resolution strategy when converting entries to mapping')
    parser.add_argument('--output', '-o', help='Output file (defaults to stdout)')

    args = parser.parse_args()

    with open(args.file, 'r') as f:
        text = f.read()

    if args.entries:
        doc = parse_nyml(text, as_entries=True)
        out = json.dumps(doc, indent=2)
    else:
        # Use mapping directly
        # For backward compatibility parse_nyml(text) returns mapping (last wins)
        # If user wants ``all`` strategy they can request entries then to_mapping
        if args.strategy == 'last':
            mapping = parse_nyml(text)
        else:
            doc = parse_nyml(text, as_entries=True)
            mapping = to_mapping(doc, strategy=args.strategy)
        out = json.dumps(mapping, indent=2)

    if args.output:
        with open(args.output, 'w') as out_f:
            out_f.write(out)
    else:
        print(out)


if __name__ == '__main__':
    main()

File: examples/nyml-cli.js
#!/usr/bin/env node

// NYML CLI - Node.js implementation
// Usage:
//  node nyml-cli.js [--entries] [--strategy last|first|all] [-o OUTPUT] FILE

const fs = require('fs');
const path = require('path');
const parser = require(path.join(__dirname, '..', 'parsers', 'javascript', 'nyml-parser.js'));

function usage() {
  console.error('Usage: nyml-cli.js [--entries] [--strategy last|first|all] [-o OUTPUT] FILE');
  process.exit(2);
}

const argv = process.argv.slice(2);
let entries = false;
let strategy = 'last';
let output = null;
let file = null;

for (let i = 0; i < argv.length; i++) {
  const a = argv[i];
  if (a === '--entries' || a === '-e') {
    entries = true;
  } else if (a.startsWith('--strategy=')) {
    strategy = a.split('=')[1] || 'last';
  } else if (a === '--strategy' || a === '-s') {
    strategy = argv[++i];
  } else if (a === '--output' || a === '-o') {
    output = argv[++i];
  } else if (a === '--help' || a === '-h') {
    usage();
  } else if (!file) {
    file = a;
  } else {
    usage();
  }
}

if (!file) usage();

const content = fs.readFileSync(file, 'utf8');
let outObj;
if (entries) {
  outObj = parser.parseNyml(content, { asEntries: true });
} else {
  if (strategy === 'last') {
    outObj = parser.parseNyml(content, {});
  } else {
    const doc = parser.parseNyml(content, { asEntries: true });
    outObj = parser.toMapping(doc, strategy);
  }
}
const outStr = JSON.stringify(outObj, null, 2);
if (output) fs.writeFileSync(output, outStr);
else console.log(outStr);
